### PR TITLE
ChatGPT Agent

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Tracking issue
https://github.com/flyteorg/flyte/issues/3936

## Why are the changes needed?
We want to support chatgpt agent and provide example for users to develop sync agent.

## What changes were proposed in this pull request?
- add openai-chatgpt plugin

### Setup process
0. Set up my local single binary environment. (use `pyflyte serve agent`)
1. Build a docker image with sync agent proto and chatgpt agent's gitsha.
2. Test the chatgpt task in both local and remote scenarios.

```yaml
tasks:
  task-plugins:
    enabled-plugins:
      - container
      - sidecar
      - K8S-ARRAY
      - agent-service
      
    default-for-task-types:
      chatgpt: agent-service
      custom_task: agent-service
      container: container
      container_array: K8S-ARRAY

plugins:
  agent-service:
    supportedTaskTypes:
      - sensor
      - spark
      - default_task
      - custom_task
      - chatgpt
      - sensor
      - airflow
    defaultAgent:
      endpoint: "dns:///localhost:8000"
      insecure: true
      timeouts:
        GetTask: 100s
      defaultTimeout: 100s
```

```Dockerfile
FROM python:3.9-slim-buster
USER root
WORKDIR /root
ENV PYTHONPATH /root
RUN apt-get update && apt-get install build-essential -y
RUN apt-get install git -y

RUN pip install -U git+https://github.com/flyteorg/flytekit.git@c9972bf023d10bef3aefa9fa6763c6b35aec5693
RUN pip install -U git+https://github.com/flyteorg/flytekit.git@c9972bf023d10bef3aefa9fa6763c6b35aec5693#subdirectory=plugins/flytekit-openai
```
```python
from flytekit import task, workflow
from flytekitplugins.chatgpt import ChatGPTTask

chatgpt_new_job = ChatGPTTask(
            name="chatgpt",
            # openai_organization="org-NayNG68kGnVXMJ8Ak4PMgQv7",
            openai_organization="xxx",
            chatgpt_config={
                    "model": "gpt-3.5-turbo",
                    "temperature": 0.7,
            },
        )

@workflow
def wf() -> str:
    message = chatgpt_new_job(message="Hi")
    return t1(s=message)

@task
def t1(s: str) -> str:
    s = "Repsonse: " + s
    return s

if __name__ == "__main__":
    print(wf())
```
#### Local Scenario
```bash
python chatgpt_chain.py
```
image
#### Remote Scenario
```bash
pyflyte run --remote --image futureoutlier/flytekit:chatgpt-0301 chatgpt_chain.py wf --message hi
```
image
image
image
image

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flytekit/pull/1822
https://github.com/flyteorg/flytekit/pull/2085

### CODE REVIEW

File: plugins/flytekit-openai/setup.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from setuptools import setup
+
+PLUGIN_NAME = "chatgpt"
+
+microlib_name = f"flytekitplugins-{PLUGIN_NAME}"
+
+plugin_requires = ["flytekit>=1.10.7", "openai>=1.12.0", "flyteidl>=1.11.0b0"]
```
Comment (Future-Outlier) [inline plugins/flytekit-openai/setup.py:7]: Should we cut a beta release of flytekit?
Since SycnAgentBase needs to be ">1.10.7" version.

Comment (Future-Outlier) [reply plugins/flytekit-openai/setup.py:7]: cc @pingsutw

Comment (pingsutw) [reply plugins/flytekit-openai/setup.py:7]: not necessary now. you can use flytekit>1.0.7. CI will always install flytekit from master branch

**Comment by Future-Outlier** [comment]
local failure case
image

remote failure case
image

File: plugins/flytekit-openai/setup.py (hunk 1 above)
Comment (Future-Outlier) [reply plugins/flytekit-openai/setup.py]: fixed, thx

File: plugins/flytekit-openai/flytekitplugins/chatgpt/agent.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        client = openai.AsyncOpenAI(
+            organization=custom["openai_organization"],
+            api_key=get_agent_secret(secret_key=OPENAI_API_KEY),
+        )
+
+        logger = logging.getLogger("httpx")
+        logger.setLevel(logging.WARNING)
+
+        try:
+            completion = await asyncio.wait_for(
```
Comment (pingsutw) [inline plugins/flytekit-openai/flytekitplugins/chatgpt/agent.py:47]: you can remove try/catch here. agent service will catch the error

File: plugins/flytekit-openai/flytekitplugins/chatgpt/agent.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+        custom = task_template.custom
+        custom["chatgpt_config"]["messages"] = [{"role": "user", "content": message}]
+        client = openai.AsyncOpenAI(
+            organization=custom["openai_organization"],
+            api_key=get_agent_secret(secret_key=OPENAI_API_KEY),
+        )
+
+        logger = logging.getLogger("httpx")
+        logger.setLevel(logging.WARNING)
```
Comment (pingsutw) [inline plugins/flytekit-openai/flytekitplugins/chatgpt/agent.py:44]: why do we need this?

File: plugins/flytekit-openai/setup.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    license="apache2",
+    python_requires=">=3.8",
+    classifiers=[
+        "Intended Audience :: Science/Research",
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: Apache Software License",
+        "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
+        "Programming Language :: Python :: 3.11",
```
Comment (pingsutw) [inline plugins/flytekit-openai/setup.py:29]:
```suggestion
        "Programming Language :: Python :: 3.11",
        "Programming Language :: Python :: 3.12",
```

File: plugins/flytekit-openai/flytekitplugins/chatgpt/agent.py (hunk 3 above)
Comment (Future-Outlier) [reply plugins/flytekit-openai/flytekitplugins/chatgpt/agent.py:44]: To hide logs when sending requests to openai client.

File: Dockerfile.agent [hunk 5]
```diff
@@ -11,7 +11,10 @@ RUN pip install prometheus-client grpcio-health-checking
 RUN pip install --no-cache-dir -U flytekit==$VERSION \
   flytekitplugins-airflow==$VERSION \
   flytekitplugins-bigquery==$VERSION \
+  flytekitplugins-chatgpt==$VERSION \
+  flytekitplugins-mmcloud==$VERSION \
   flytekitplugins-snowflake==$VERSION \
+  flytekitplugins-spark==$VERSION \
```
Comment (pingsutw) [inline Dockerfile.agent:17]:
```suggestion
  flytekitplugins-snowflake==$VERSION \
```

**Pull request merged**